# Two mathematicians
This riddle I first heard in 2002 from Stefan Bergström and I'm not sure I solved it then. If I did, I don't remember the answer. This notebook solves it!

## The riddle
Two mathematicians, let's call them S and P know the sum (S) and product (P) of two positive integers between 3 and 100. They are both aware of the setup of the riddle and are perfectly logical in everything they do. The following dialog occurs.

- S: "I know that you don't know which the two integers are."
- P: "Then I know which they are!"
- S: "Then so do I!"

Which are the integers?

## Solution - statement 1
For P not to be able to work out the numbers from the product, they can not be two primes. For example, if the product was 21, then the two integers must be 3 and 7, as no other combination ofpositive integers (between 3 and 100) can create this product. For S to know this, the sum of integers, in this case 10, should not be able to form by two primes. If the sum is 10, the integers can be 3,7 or 4,6 or 5,5. If they were 4 and 6, then the product would be 24 and it would be impossible for P to know if the two integers were 6 and 4 or 3 and 12.
However, since S only knows that the sum is 10, the integers *could* be 3 and 7, or 5 and 5, and the statement "I *know* that you don't know..." could not be stated.

Thus, let's start by finding all possible candidates given by this first statement.

In [74]:
N_MIN = 3
N_MAX = 100

import numpy as np
import pandas as pd
all_possible_combos = np.array(np.meshgrid(range(N_MIN,N_MAX+1), range(N_MIN,N_MAX+1))).reshape(2, (N_MAX-N_MIN+1)**2).T
all_possible_combos = pd.DataFrame(all_possible_combos, columns=['x', 'y'])
all_possible_combos = all_possible_combos[all_possible_combos.x >= all_possible_combos.y].sort_values(by=['x', 'y']).reset_index(drop=True)
all_possible_combos['s'] = all_possible_combos.x + all_possible_combos.y
all_possible_combos['p'] = all_possible_combos.x * all_possible_combos.y
print('We have', len(all_possible_combos), 'possible combinations to start with')
# all_possible_combos

We have 4851 possible combinations to start with


In [75]:
def is_prime(n):
    assert (n <= 101) and (n >= 2), "This function only works for integers between 3 and 100. Aborting"
    return(n in [2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47, 53, 59, 61, 67, 71, 73, 79, 83, 89, 97, 101])

# is_prime(-1)
# is_prime(10)
# is_prime(11)

# Determine if both x and y are prime
all_possible_combos['x_and_y_is_prime'] = all_possible_combos.apply(lambda r: is_prime(r.x) and is_prime(r.y),1)

# Find all sums which contain one instance of x and y being primes
temp = all_possible_combos.loc[all_possible_combos.x_and_y_is_prime, 's'].unique()

# Remove these sums
all_possible_combos = all_possible_combos[all_possible_combos.s.isin(temp)].reset_index(drop=True)
print('We have now have', len(all_possible_combos), 'possible combinations left')
all_possible_combos

We have now have 2393 possible combinations left


x   y    s     p  x_and_y_is_prime
0       3   3    6     9              True
1       4   4    8    16             False
2       5   3    8    15              True
3       5   5   10    25              True
4       6   4   10    24             False
...   ...  ..  ...   ...               ...
2388  100  76  176  7600             False
2389  100  78  178  7800             False
2390  100  80  180  8000             False
2391  100  86  186  8600             False
2392  100  94  194  9400             False

[2393 rows x 5 columns]

## Solution - statement 2
Now we got 2,393 possible combinations left, so let's analyze the second statement: "Then I know which they are".

For P to be able to say this, the product he knows must be unique. Thus let's remove all integer combinations that do not give rise to a unique product.

In [76]:
temp = all_possible_combos[['p', 'x']].groupby('p').count().reset_index()
all_possible_combos = all_possible_combos[all_possible_combos.p.isin(temp.p[temp.x == 1])].reset_index(drop=True)
print('We have now have', len(all_possible_combos), 'possible combinations left')
all_possible_combos

We have now have 1309 possible combinations left


x   y    s     p  x_and_y_is_prime
0       3   3    6     9              True
1       4   4    8    16             False
2       5   3    8    15              True
3       5   5   10    25              True
4       6   4   10    24             False
...   ...  ..  ...   ...               ...
1304  100  76  176  7600             False
1305  100  78  178  7800             False
1306  100  80  180  8000             False
1307  100  86  186  8600             False
1308  100  94  194  9400             False

[1309 rows x 5 columns]

## Solution - statement 3
OK, so P has figured it out, but we haven't. We still have 1,309 possible combinations. Let's analyze the third statement which is S saying "Then I know too!".

In order for him to say that, the sum must be unique among the remaining possible candidates. Let's remove all non-unique sums and see what we get.

In [77]:
temp = all_possible_combos[['s', 'x']].groupby('s').count().reset_index()
all_possible_combos = all_possible_combos[all_possible_combos.s.isin(temp.s[temp.x == 1])]
print('We have now have', len(all_possible_combos), 'possible combination left!')
all_possible_combos

We have now have 1 possible combination left!


x  y  s  p  x_and_y_is_prime
0  3  3  6  9              True

## Answer
This is the answer! The two integers are 3 and 3! I'm actually a bit surprised that I solved it :-)